cmd/clusterctl/client/cluster/topology.go

```diff
@@ -0,0 +1,613 @@
+		}
+		if err := json.Unmarshal(originalWithModifiedJSON, &cluster.Object); err != nil {
+			return errors.Wrap(err, "failed to convert modified json to object")
+		}
+	}
+	return nil
+}
+
+// defaultAndValidate runs the defaulting and validation webhooks on the
+// ClusterClass and Cluster objects in the input.
+// Nb. Perform ValidateUpdate only if the object is already in the cluster. In all other cases,
+// ValidateCreate is performed.
+func (t *topologyClient) defaultAndValidate(in *DryRunInput, apiReader client.Reader) error {
+	// Enable the ClusterTopology feature gate so that the defaulter and validators do not complain.
+	// Note: We don't ned to disable it later because the CLI is short lived.
+	if err := feature.Gates.(featuregate.MutableFeatureGate).Set(fmt.Sprintf("%s=%v", feature.ClusterTopology, true)); err != nil {
+		return errors.Wrapf(err, "failed to enable %s feature gate", feature.ClusterTopology)
+	}
+
+	filteredObjs := filterObjects(
+		in.Objs,
+		clusterv1.GroupVersion.WithKind("ClusterClass"),
+		clusterv1.GroupVersion.WithKind("Cluster"),
+	)
+	objs := []client.Object{}
+	for _, o := range filteredObjs {
+		objs = append(objs, o)
+	}
+	webhookClient := dryrun.NewClient(apiReader, objs)
+
+	// Run defaulting and validation on ClusterClasses.
+	ccWebhook := &webhooks.ClusterClass{Client: webhookClient}
+	if err := t.defaultAndValidateObjs(
+		getClusterClasses(in.Objs),
+		&clusterv1.ClusterClass{},
+		ccWebhook,
+		ccWebhook,
+		apiReader,
+	); err != nil {
+		return errors.Wrap(err, "failed to run defaulting and validation on ClusterClasses")
+	}
+
+	filteredObjs = filterObjects(
+		in.Objs,
+		clusterv1.GroupVersion.WithKind("Cluster"),
+	)
+	objs = []client.Object{}
+	for _, o := range filteredObjs {
+		objs = append(objs, o)
+	}
+	webhookClient = dryrun.NewClient(apiReader, objs)
+
+	// Run defaulting and validation on Clusters.
+	clusterWebhook := &webhooks.Cluster{Client: webhookClient}
+	if err := t.defaultAndValidateObjs(
+		getClusters(in.Objs),
+		&clusterv1.Cluster{},
+		clusterWebhook,
+		clusterWebhook,
+		apiReader,
+	); err != nil {
+		return errors.Wrap(err, "failed to run defaulting and validation on Clusters")
+	}
+
+	return nil
+}
+
+func (t *topologyClient) defaultAndValidateObjs(objs []*unstructured.Unstructured, o client.Object, defaulter crwebhook.CustomDefaulter, validator crwebhook.CustomValidator, apiReader client.Reader) error {
+	ctx := context.TODO()
+	for _, obj := range objs {
+		object := o.DeepCopyObject().(client.Object)
+		if err := localScheme.Convert(obj, object, nil); err != nil {
+			return errors.Wrapf(err, "failed to convert object to %s", obj.GetKind())
+		}
+
+		// Perform Defaulting
+		if err := defaulter.Default(ctx, object); err != nil {
+			return errors.Wrapf(err, "failed defaulting of %s %s/%s", obj.GroupVersionKind().String(), obj.GetNamespace(), obj.GetName())
+		}
+
+		// Perform Validation
+		performCreateValidation := true
+		if apiReader != nil {
```

Comment:
are there cases where the apiReader is not / should be not set? If there aren't I wouldn't tolerate that case.

Reply:
Okay I'll check again on the next review. I thought I went through the usages and didn't fine one, but I probably missed it.